orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,94 @@
+package cluster
+
+import (
+	"github.com/hyperledger/fabric-protos-go/common"
+	"github.com/hyperledger/fabric/bccsp"
+)
+
+//go:generate mockery -dir . -name BlockSource -case underscore -output mocks/
+
+type BlockSource interface {
+	PullBlock(seq uint64) *common.Block
+	UpdateEndpoints(endpoints []EndpointCriteria)
+}
+
+type BlockFetcher struct {
+	source                 BlockSource
+	Endpoints              []EndpointCriteria
+	suspects               []string // a list of bft suepected notes, for optimisation later can be converted to a set.
+	endpointToPull         string
+	endpointUnderSuspicion bool
+}
+
+func (bf *BlockFetcher) isSuspect(suspect string) bool {
+	for _, r := range bf.suspects {
+		if suspect == r {
+			return true
+		}
+	}
+	return false
+}
+
+func (bf *BlockFetcher) chooseEndpoint(ec EndpointCriteria) {
+	// set source of BlockPuller
+	bf.endpointToPull = ec.Endpoint
+	bf.source.UpdateEndpoints([]EndpointCriteria{ec})
+}
+
+func (bf *BlockFetcher) PullBlock(seq uint64) *common.Block {
+
+	for _, endpoint := range bf.Endpoints {
+
+		// check for the endpoint in bft suspects list
+		if bf.isSuspect(endpoint.Endpoint) {
+			continue
+		}
+
+		// set source of BlockPuller
+		bf.chooseEndpoint(endpoint)
+		block := bf.source.PullBlock(seq)
+
+		switch block {
+		case nil:
+			// check suspicion by probing other nodes
+			if bf.suspectCurrentEndpoint() {
+				bf.suspects = append(bf.suspects, endpoint.Endpoint)
+			}
+
+		default:
+			// check if it is time to switch to another node
+			if bf.endpointUnderSuspicion {
+				bf.suspects = append(bf.suspects, endpoint.Endpoint)
```

Comment:
you are risking that the `suspects` will grow indefinitely no? Since it's not a set, but an array

Reply:
Agree. Need a set.